### PR TITLE
datatrails: only store bookmark state instead of the entire trail step history

### PR DESCRIPTION
**Pull request by darrenjaneczek**
**What is this feature?**

When storing bookmarks, only the bookmark URL state is retained, instead of the entire trail history.

**Why do we need this feature?**

We were questioning whether or not it made sense for the entire trail history to be stored.

**Special notes for your reviewer:**

Questions:
- Do we want to recover bookmarks from the old format?

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/trails/shared.ts [hunk 1]
```diff
@@ -32,7 +32,8 @@ export const trailDS = { uid: VAR_DATASOURCE_EXPR };
 
 // Local storage keys
 export const RECENT_TRAILS_KEY = 'grafana.trails.recent';
-export const BOOKMARKED_TRAILS_KEY = 'grafana.trails.bookmarks';
+
+export const BOOKMARKED_STATES_KEY = 'grafana.trails.bookmarkedStates';
```
Comment (torkelo) [inline public/app/features/trails/shared.ts:36]: any way we can keep the old key and just detect / migrate the old ones?

File: public/app/features/trails/TrailStore/TrailStore.ts [hunk 2]
```diff
@@ -112,8 +149,8 @@ export class TrailStore {
   }
 
   load() {
-    this._recent = this._loadFromStorage(RECENT_TRAILS_KEY);
-    this._bookmarks = this._loadFromStorage(BOOKMARKED_TRAILS_KEY);
+    this._recent = this._loadRecentTrailsFromStorage();
+    this._bookmarkedStates = this._loadBookmarkedStatesFromStorage();
```
Comment (torkelo) [inline public/app/features/trails/TrailStore/TrailStore.ts:153]: not sure you need all the "States" suffixes everywhere (it feels unclear either way, what is a bookmark state vs just a bookmark?)

File: public/app/features/trails/TrailStore/TrailStore.ts [hunk 3]
```diff
@@ -23,9 +24,14 @@ export interface SerializedTrail {
   createdAt?: number;
 }
 
+export interface SerializedBookmarkState {
```
Comment (torkelo) [inline public/app/features/trails/TrailStore/TrailStore.ts:27]:
```suggestion
export interface DataTrailBookmark {
```

Not sure the sure the Seralized and Sate prefix/suffix adds much

**Review by torkelo — COMMENTED** [review]
Looks good!

File: public/app/features/trails/shared.ts (hunk 1 above)
Comment (darrenjaneczek) [reply public/app/features/trails/shared.ts]: Done!
Simplifies much.

**Comment by diegommm** [comment]
Hi! Greetings from the Release Guild. Just letting you know I'm updating your PR's milestone to [11.2.x](https://github.com/grafana/grafana/milestone/529). If you need to see your changes in a previous version, please consider adding a backport label.

Thank you!

File: public/app/features/trails/TrailStore/TrailStore.test.ts [hunk 4]
```diff
@@ -635,44 +671,122 @@ describe('TrailStore', () => {
       expect(store.recent.length).toBe(0);
     });
 
-    it('should accurately load bookmarked trails', () => {
+    it('should accurately load legacy bookmark', () => {
```
Comment (bohandley) [inline public/app/features/trails/TrailStore/TrailStore.test.ts:674]: Nice! This looks good. I was looking for the migration of old bookmarks.

**Review by bohandley — APPROVED** [review]
All looks good to me. For testing purposes, I made some bookmarks in the legacy version, switched to the new branch, they loaded fine. The new bookmark saves a metric and upon loading it again starts a trail at the second step, which looked odd at first but makes sense. A user bookmarked a metric and upon returning to the bookmark they had already started a trail.

Also, I added a panel to KPI dashboard about bookmarks. There aren't a lot of users interacting with them but as we're changing this, we should keep and eye on the usage
https://bi.grafana-ops.net/d/adkmlvq4wpjpcb/explore-metrics-kpi-wip?tab=queries&editPanel=11

**Comment by darrenjaneczek** [comment]
@bohandley 
> The new bookmark saves a metric and upon loading it again starts a trail at the second step, which looked odd at first but makes sense.

I have mixed feelings about this as well. I wouldn't be bothered if in the future, reloaded bookmarks only had one "metric orange" step, instead of the superfluous no-metric-selected step.

**Pull request merged**